// santa-animations/src/definitions/animation/backgroundSmoothScrollEffects/bgCloseUp.js
'use strict'
const name = 'BgCloseUp'
const properties = {
    hideOnStart: false,
    getMaxTravel(elementMeasure, viewPortHeight) {
        return viewPortHeight + elementMeasure.height
    },
    groups: ['animation', 'background'],
    schema: {
        duration: {
            type: 'number',
            min: 0,
            default: 0
        },
        delay: {
            type: 'number',
            min: 0,
            default: 0
        },
        componentHeight: {
            type: 'number',
            min: 0
        },
        viewPortHeight: {
            type: 'number',
            min: 0
        },
        perspectiveParent: {
            type: 'element'
        }
    }
}

function register({
    factory
}) {
    /**
     * @param {HTMLElement|HTMLElement[]} elements DOM elements to animate
     * @param {Number} [duration]
     * @param {Number} [delay]
     * @param {Object} [params] Timeline optional parameters.
     * @returns {TimelineMax}
     */
    function animation(elements, duration, delay, {
        viewPortHeight,
        componentHeight,
        perspectiveParent,
        ...params
    } = {}) {
        const sequence = factory.sequence(params)

        sequence.add([
            ...elements.map(element => factory.animate('BaseBgZoom', element.children, duration, delay, {
                viewPortHeight,
                componentHeight,
                perspectiveParent,
                in: {
                    start: 0,
                    end: 1,
                    scale: 5,
                    ease: 'none'
                }
            })),
            factory.animate('BaseBgFade', elements, duration, delay, {
                out: {
                    start: 0.55,
                    end: 0.95,
                    opacity: 0,
                    ease: 'none'
                }
            })
        ])

        return sequence.get()
    }

    factory.registerAnimation(name, animation, properties)
}

module.exports = {
    name,
    properties,
    register
}
