// santa-animations/src/definitions/animation/backgroundSmoothScrollEffects/bgFadeBack.js
'use strict'
const name = 'BgFadeBack'
const properties = {
    hideOnStart: false,
    getMaxTravel(elementMeasure, viewPortHeight) {
        return viewPortHeight + elementMeasure.height
    },
    groups: ['animation', 'background'],
    schema: {
        duration: {
            type: 'number',
            min: 0,
            default: 0
        },
        delay: {
            type: 'number',
            min: 0,
            default: 0
        }
    }
}

function register({
    factory
}) {
    /**
     * @param {HTMLElement|HTMLElement[]} elements DOM elements to animate
     * @param {Number} [duration]
     * @param {Number} [delay]
     * @param {Object} [params] Timeline optional parameters.
     * @returns {TimelineMax}
     */
    function animation(elements, duration, delay, params = {}) {
        const sequence = factory.sequence(params)

        sequence.add([
            factory.animate('BaseBgFade', elements, duration, delay, {
                out: {
                    start: 0.6,
                    end: 1,
                    opacity: 0,
                    ease: 'none'
                }
            }),
            factory.animate('BaseBgScale', elements, duration, delay, {
                out: {
                    start: 0.6,
                    end: 0.85,
                    scaleX: 0.7,
                    scaleY: 0.7,
                    ease: 'sine.out'
                }
            })
        ])

        return sequence.get()
    }

    factory.registerAnimation(name, animation, properties)
}

module.exports = {
    name,
    properties,
    register
}
