// santa-animations/src/definitions/animation/backgroundSmoothScrollEffects/bgPullBack.js
'use strict'
const name = 'BgPullBack'
const properties = {
    hideOnStart: false,
    getMaxTravel(elementMeasure, viewPortHeight) {
        return viewPortHeight + elementMeasure.height
    },
    groups: ['animation', 'background'],
    schema: {
        duration: {
            type: 'number',
            min: 0,
            default: 0
        },
        delay: {
            type: 'number',
            min: 0,
            default: 0
        },
        perspectiveParent: {
            type: 'element'
        },
        componentHeight: {
            type: 'number',
            min: 0
        },
        viewPortHeight: {
            type: 'number',
            min: 0
        },
        speedFactor: {
            type: 'number',
            default: 1
        }
    }
}

function register({
    factory
}) {
    /**
     * @param {HTMLElement|HTMLElement[]} elements DOM elements to animate
     * @param {Number} [duration]
     * @param {Number} [delay]
     * @param {Object} [params] Timeline optional parameters.
     * @returns {TimelineMax}
     */
    function animation(elements, duration, delay, {
        perspectiveParent,
        componentHeight,
        viewPortHeight,
        //speedFactor = properties.schema.speedFactor.default,
        ...params
    } = {}) {
        const sequence = factory.sequence(params)
        const end = Math.min(componentHeight / (viewPortHeight + componentHeight), 1)
        sequence.add([
            factory.animate('BaseBgZoom', elements, duration, delay, {
                viewPortHeight,
                componentHeight,
                perspectiveParent,
                speedFactor: 0,
                out: {
                    start: 0,
                    end,
                    scale: 2,
                    ease: 'none'
                }
            }),
            factory.animate('BaseBgPositionY', elements, duration, delay, {
                from: -componentHeight / 6,
                to: 0,
                start: 0,
                end,
                ease: 'none'
            })
        ])

        return sequence.get()
    }

    factory.registerAnimation(name, animation, properties)
}

module.exports = {
    name,
    properties,
    register
}
