// santa-animations/src/definitions/animation/backgroundSmoothScrollEffects/bgZoomOut.js
'use strict'
const name = 'BgZoomOut'
const properties = {
    hideOnStart: false,
    getMaxTravel(elementMeasure, viewPortHeight) {
        return viewPortHeight + elementMeasure.height
    },
    groups: ['animation', 'background'],
    schema: {
        duration: {
            type: 'number',
            min: 0,
            default: 0
        },
        delay: {
            type: 'number',
            min: 0,
            default: 0
        },
        componentHeight: {
            type: 'number',
            min: 0
        },
        viewPortHeight: {
            type: 'number',
            min: 0
        },
        perspectiveParent: {
            type: 'element'
        }
    }
}

function register({
    factory
}) {
    /**
     * @param {HTMLElement|HTMLElement[]} elements DOM elements to animate
     * @param {Number} [duration]
     * @param {Number} [delay]
     * @param {Object} [params] Timeline optional parameters.
     * @returns {TimelineMax}
     */
    function animation(elements, duration, delay, {
        perspectiveParent,
        viewPortHeight,
        componentHeight,
        ...params
    } = {}) {
        const sequence = factory.sequence(params)

        sequence.add(
            factory.animate('BaseBgZoom', elements, duration, delay, {
                viewPortHeight,
                componentHeight,
                perspectiveParent,
                out: {
                    start: 0,
                    end: 1,
                    scale: 1.3,
                    ease: 'sine.inOut'
                }
            })
        )

        return sequence.get()
    }

    factory.registerAnimation(name, animation, properties)
}

module.exports = {
    name,
    properties,
    register
}
