// @wix/editor-elements-corvid-utils/dist/props-factories/animations/constants.js
export var effectDefaultOptions = {
    arc: {
        duration: 1.2,
        delay: 0,
        direction: 'left',
    },
    bounce: {
        duration: 1.2,
        delay: 0,
        direction: 'topLeft',
        bounce: 'medium',
    },
    puff: {
        duration: 1.2,
        delay: 0,
    },
    zoom: {
        duration: 1.2,
        delay: 0,
    },
    fade: {
        duration: 1.2,
        delay: 0,
    },
    flip: {
        duration: 1.2,
        delay: 0,
        direction: 'left',
    },
    float: {
        duration: 1.2,
        delay: 0,
        direction: 'left',
    },
    fly: {
        duration: 1.2,
        delay: 0,
        direction: 'left',
    },
    fold: {
        duration: 1.2,
        delay: 0,
        direction: 'left',
    },
    glide: {
        duration: 1.2,
        delay: 0,
        angle: 0,
        distance: 0,
    },
    roll: {
        duration: 1.2,
        delay: 0,
        direction: 'left',
    },
    slide: {
        duration: 1.2,
        delay: 0,
        direction: 'left',
    },
    spin: {
        duration: 1.2,
        delay: 0,
        direction: 'cw',
        cycles: 5,
    },
    turn: {
        duration: 1.2,
        delay: 0,
        direction: 'right',
    },
};
export var EFFECTS = {
    HIDE: {
        suffix: 'out',
        deprecatedValues: [
            'ArcOut',
            'BounceOut',
            'CollapseOut',
            'Conceal',
            'FadeOut',
            'FlipOut',
            'FloatOut',
            'FlyOut',
            'FoldOut',
            'GlideOut',
            'PopOut',
            'SlideOut',
            'SpinOut',
            'TurnOut',
        ],
    },
    SHOW: {
        suffix: 'in',
        deprecatedValues: [
            'ArcIn',
            'BounceIn',
            'DropIn',
            'ExpandIn',
            'FadeIn',
            'FlipIn',
            'FloatIn',
            'FlyIn',
            'FoldIn',
            'GlideIn',
            'Reveal',
            'SlideIn',
            'SpinIn',
            'TurnIn',
        ],
    },
};
var capitalize = function(str) {
    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
};
export var effectInfoLink = function(functionName) {
    return "https://www.wix.com/code/reference/$w.HiddenMixin.html#" + functionName;
};
export var effectOptionsForEffectInfoLink = function(effectName) {
    return "https://www.wix.com/code/reference/$w.EffectOptions.html#" + capitalize(effectName) + "EffectOptions";
};
