// @wix/advanced-seo-utils/dist/es/src/renderer-api.js
import {
    resolveWithPatterns
} from './private/renderer/resolve-with-patterns';
import {
    filterPageLevelTags
} from './private/tags/filters/filter-page-level-tags';
import {
    convertSeoModel
} from './private/converters/convert-seo-model';
import {
    convertSiteLevelMetaTags
} from './private/converters/convert-site-level-meta-tags';
import {
    adapter as staticPageAdapterV2
} from './private/item-types/static-page-v2';
import {
    STATIC_PAGE_V2
} from './private/item-types/static-page-v2/pattern-static-page-v2';
import {
    enrichWithSiteLevelData
} from './private/adapters/utils';
import {
    getUserPatternByItemType
} from './private/patterns/get-user-pattern-by-item-type';
import {
    getContext
} from './private/tags/context/get-context';
import {
    getDefaultItemPayload
} from './private/patterns/get-default-item';
export {
    getTitle,
    getLinks,
    getMetaTags,
    getSchemas,
}
from './private/renderer/getters';
export {
    setExternalRouter
}
from './private/renderer/setters';
export {
    getInitialTags,
    getDefaultItemPayload
};

function getInitialTags(_a) {
    var _b = _a.siteLevelSeoData,
        siteLevelSeoData = _b === void 0 ? {} : _b,
        pageLevelSeoData = _a.pageLevelSeoData,
        _c = _a.dynamicPageData,
        dynamicPageData = _c === void 0 ? [] : _c;
    var siteLevelMetaTags = filterPageLevelTags(convertSiteLevelMetaTags(siteLevelSeoData.metaTags));
    var defaultStaticPatternSchema = STATIC_PAGE_V2;
    var userPatternSchema = getUserPatternByItemType(siteLevelSeoData.userPatterns);
    var adapter = enrichWithSiteLevelData(staticPageAdapterV2);
    var context = getContext(siteLevelSeoData, pageLevelSeoData, userPatternSchema);
    var contextWithItemData = {
        context: context
    };
    var editorPanelLegacyData = adapter.getLegacySeoBlob({
        context: context
    });
    var editorPanelData = convertSeoModel(pageLevelSeoData.advancedSeoData);
    var payload = [
        siteLevelMetaTags,
        defaultStaticPatternSchema,
        userPatternSchema,
        editorPanelLegacyData,
        editorPanelData,
        dynamicPageData,
    ];
    var adaptedContext = adapter.getData(contextWithItemData);
    return resolveWithPatterns(payload, adaptedContext);
}
