// image-client-api/dist/imageClientApi.js
! function(e, t) {
    "object" == typeof exports && "object" == typeof module ? module.exports = t() : "function" == typeof define && define.amd ? define([], t) : "object" == typeof exports ? exports.imageClientApi = t() : e.imageClientApi = t()
}(this, (function() {
    return function(e) {
        var t = {};

        function i(a) {
            if (t[a]) return t[a].exports;
            var r = t[a] = {
                i: a,
                l: !1,
                exports: {}
            };
            return e[a].call(r.exports, r, r.exports, i), r.l = !0, r.exports
        }
        return i.m = e, i.c = t, i.d = function(e, t, a) {
            i.o(e, t) || Object.defineProperty(e, t, {
                enumerable: !0,
                get: a
            })
        }, i.r = function(e) {
            "undefined" != typeof Symbol && Symbol.toStringTag && Object.defineProperty(e, Symbol.toStringTag, {
                value: "Module"
            }), Object.defineProperty(e, "__esModule", {
                value: !0
            })
        }, i.t = function(e, t) {
            if (1 & t && (e = i(e)), 8 & t) return e;
            if (4 & t && "object" == typeof e && e && e.__esModule) return e;
            var a = Object.create(null);
            if (i.r(a), Object.defineProperty(a, "default", {
                    enumerable: !0,
                    value: e
                }), 2 & t && "string" != typeof e)
                for (var r in e) i.d(a, r, function(t) {
                    return e[t]
                }.bind(null, r));
            return a
        }, i.n = function(e) {
            var t = e && e.__esModule ? function() {
                return e.default
            } : function() {
                return e
            };
            return i.d(t, "a", t), t
        }, i.o = function(e, t) {
            return Object.prototype.hasOwnProperty.call(e, t)
        }, i.p = "", i(i.s = 11)
    }([function(e, t, i) {
        "use strict";
        var a = {
                JPG: "jpg",
                JPEG: "jpeg",
                JPE: "jpe",
                PNG: "png",
                WEBP: "webp",
                WIX_ICO_MP: "wix_ico_mp",
                WIX_MP: "wix_mp",
                GIF: "gif",
                SVG: "svg",
                UNRECOGNIZED: "unrecognized"
            },
            r = [a.JPG, a.JPEG, a.JPE, a.PNG, a.GIF, a.WEBP];
        e.exports = {
            alignTypes: {
                CENTER: "center",
                TOP: "top",
                TOP_LEFT: "top_left",
                TOP_RIGHT: "top_right",
                BOTTOM: "bottom",
                BOTTOM_LEFT: "bottom_left",
                BOTTOM_RIGHT: "bottom_right",
                LEFT: "left",
                RIGHT: "right"
            },
            alignTypesMap: {
                center: "c",
                top: "t",
                top_left: "tl",
                top_right: "tr",
                bottom: "b",
                bottom_left: "bl",
                bottom_right: "br",
                left: "l",
                right: "r"
            },
            transformTypes: {
                FIT: "fit",
                FILL: "fill",
                FILL_FOCAL: "fill_focal",
                CROP: "crop",
                LEGACY_CROP: "legacy_crop",
                LEGACY_FILL: "legacy_fill"
            },
            fittingTypes: {
                SCALE_TO_FILL: "fill",
                SCALE_TO_FIT: "fit",
                STRETCH: "stretch",
                ORIGINAL_SIZE: "original_size",
                TILE: "tile",
                TILE_HORIZONTAL: "tile_horizontal",
                TILE_VERTICAL: "tile_vertical",
                FIT_AND_TILE: "fit_and_tile",
                LEGACY_STRIP_TILE: "legacy_strip_tile",
                LEGACY_STRIP_TILE_HORIZONTAL: "legacy_strip_tile_horizontal",
                LEGACY_STRIP_TILE_VERTICAL: "legacy_strip_tile_vertical",
                LEGACY_STRIP_SCALE_TO_FILL: "legacy_strip_fill",
                LEGACY_STRIP_SCALE_TO_FIT: "legacy_strip_fit",
                LEGACY_STRIP_FIT_AND_TILE: "legacy_strip_fit_and_tile",
                LEGACY_STRIP_ORIGINAL_SIZE: "legacy_strip_original_size",
                LEGACY_ORIGINAL_SIZE: "actual_size",
                LEGACY_FIT_WIDTH: "fitWidth",
                LEGACY_FIT_HEIGHT: "fitHeight",
                LEGACY_FULL: "full",
                LEGACY_BG_FIT_AND_TILE: "legacy_tile",
                LEGACY_BG_FIT_AND_TILE_HORIZONTAL: "legacy_tile_horizontal",
                LEGACY_BG_FIT_AND_TILE_VERTICAL: "legacy_tile_vertical",
                LEGACY_BG_NORMAL: "legacy_normal"
            },
            htmlTag: {
                BG: "bg",
                IMG: "img",
                SVG: "svg"
            },
            upscaleMethods: {
                AUTO: "auto",
                CLASSIC: "classic",
                SUPER: "super"
            },
            upscaleMethodsValues: {
                classic: 1,
                super: 2
            },
            defaultUSM: {
                radius: .66,
                amount: 1,
                threshold: .01
            },
            emptyData: {
                uri: "",
                css: {
                    img: {},
                    container: {}
                },
                attr: {
                    img: {},
                    container: {}
                }
            },
            imageQuality: {
                HIGH: "HIGH",
                MEDIUM: "MEDIUM",
                LOW: "LOW",
                TINY: "TINY"
            },
            imageFilters: {
                CONTRAST: "contrast",
                BRIGHTNESS: "brightness",
                SATURATION: "saturation",
                HUE: "hue",
                BLUR: "blur"
            },
            imageScaleDefaults: {
                HIGH: {
                    size: 196e4,
                    quality: 90,
                    maxUpscale: 1
                },
                MEDIUM: {
                    size: 36e4,
                    quality: 85,
                    maxUpscale: 1
                },
                LOW: {
                    size: 16e4,
                    quality: 80,
                    maxUpscale: 1.2
                },
                TINY: {
                    size: 0,
                    quality: 80,
                    maxUpscale: 1.4
                }
            },
            fileType: a,
            supportedExtensions: r,
            webp: {
                LOSSLESS: "lossless",
                LOSSY: "lossy",
                ALPHA: "alpha",
                ANIMATION: "animation"
            },
            DSKTP_MAX_BG_SITE_LEGACY_WIDTH: 1920,
            MOBILE_MAX_BG_SITE_LEGACY_WIDTH: 1e3,
            DSKTP_MAX_BG_SITE_LEGACY_HEIGHT: 1920,
            MOBILE_MAX_BG_SITE_LEGACY_HEIGHT: 1e3,
            SAFE_TRANSFORMED_AREA: 25e6,
            SUPER_UPSCALE_MODELS: [1.5, 2, 4],
            MAX_DEVICE_PIXEL_RATIO: 2,
            API_VERSION: "v1"
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(5),
            r = i(6),
            n = i(2),
            o = i(0),
            s = i(3);

        function c(e) {
            var t = [o.fileType.PNG, o.fileType.JPEG, o.fileType.JPG, o.fileType.JPE, o.fileType.WIX_ICO_MP, o.fileType.WIX_MP];
            return n.includes(t, p(e))
        }

        function h(e) {
            return n.includes(["webp"], p(e))
        }

        function u(e) {
            return /(^https?)|(^data)|(^\/\/)/.test(e)
        }

        function p(e) {
            return (/[.]([^.]+)$/.exec(e) && /[.]([^.]+)$/.exec(e)[1] || "").toLowerCase()
        }

        function l(e, t, i, a, r) {
            return r === o.transformTypes.FILL ? function(e, t, i, a) {
                return Math.max(i / e, a / t)
            }(e, t, i, a) : r === o.transformTypes.FIT ? function(e, t, i, a) {
                return Math.min(i / e, a / t)
            }(e, t, i, a) : 1
        }

        function g(e, t) {
            var i = m(e, t);
            return {
                optimizedScaleFactor: o.imageScaleDefaults[i].maxUpscale,
                upscaleMethodValue: o.upscaleMethodsValues.classic,
                forceUSM: !1
            }
        }

        function T(e, t) {
            var i = m(e, t);
            return {
                optimizedScaleFactor: o.imageScaleDefaults[i].maxUpscale,
                upscaleMethodValue: o.upscaleMethodsValues.classic,
                forceUSM: !1
            }
        }

        function f(e, t, i) {
            return {
                optimizedScaleFactor: n.last(o.SUPER_UPSCALE_MODELS),
                upscaleMethodValue: o.upscaleMethodsValues.super,
                forceUSM: !(o.SUPER_UPSCALE_MODELS.includes(i) || i > n.last(o.SUPER_UPSCALE_MODELS))
            }
        }

        function m(e, t) {
            var i = e * t;
            return i > o.imageScaleDefaults[o.imageQuality.HIGH].size ? o.imageQuality.HIGH : i > o.imageScaleDefaults[o.imageQuality.MEDIUM].size ? o.imageQuality.MEDIUM : i > o.imageScaleDefaults[o.imageQuality.LOW].size ? o.imageQuality.LOW : o.imageQuality.TINY
        }

        function _(e, t) {
            var i = Math.pow(10, t || 0);
            return (e * i / i).toFixed(parseInt(t, 10))
        }
        e.exports = {
            populateGlobalFeatureSupport: function(e) {
                var t;
                void 0 === e && (e = ""), "undefined" != typeof window && "undefined" != typeof navigator ? (t = a(navigator.userAgent), r.checkSupportByUserAgent(e, t), r.checkSupportByFeatureDetection(), s.setFeature("isObjectFitBrowser", "objectFit" in window.document.documentElement.style), s.setFeature("isMobile", t.os.phone)) : (t = a(e), r.checkSupportByUserAgent(e, t), s.setFeature("isMobile", t.os.phone))
            },
            isWEBPBrowserSupport: r.isWEBPBrowserSupport,
            isObjectFitBrowserSupport: function() {
                return s.getFeature("isObjectFitBrowser")
            },
            isMobile: function() {
                return s.getFeature("isMobile")
            },
            isImageTransformApplicable: function(e) {
                return c(e) && !u(e)
            },
            isValidRequest: function(e, t, i) {
                return i && t && !(!(a = t.id) || !a.trim() || "none" === a.toLowerCase()) && n.includes(o.fittingTypes, e);
                var a
            },
            isImageTypeSupported: c,
            isExternalUrl: u,
            isWEBP: h,
            isSEOBot: function(e) {
                return e && e.isSEOBot || !1
            },
            getFileType: function(e) {
                return function(e) {
                    return n.includes([o.fileType.JPEG, o.fileType.JPG, o.fileType.JPE], p(e))
                }(e) ? o.fileType.JPG : function(e) {
                    return n.includes(["png"], p(e))
                }(e) ? o.fileType.PNG : h(e) ? o.fileType.WEBP : o.fileType.UNRECOGNIZED
            },
            getFileExtension: p,
            getFileName: function(e, t) {
                var i = /\.([^.]*)$/;
                if ("string" == typeof t && t.length) {
                    var a = ["/", "\\", "?", "<", ">", "|", "“", ":", '"'].map(encodeURIComponent),
                        r = new RegExp("(" + a.concat(["\\.", "\\*"]).join("|") + ")", "g"),
                        s = t,
                        c = t.match(i);
                    return c && n.includes(o.supportedExtensions, c[1]) && (s = t.replace(i, "")), encodeURIComponent(s).replace(r, "_")
                }
                var h = e.match(/\/(.*?)$/);
                return (h ? h[1] : e).replace(i, "")
            },
            getAlignedRect: function(e, t, i) {
                var a, r;
                switch (i) {
                    case o.alignTypes.CENTER:
                        a = Math.max(0, (e.width - t.width) / 2), r = Math.max(0, (e.height - t.height) / 2);
                        break;
                    case o.alignTypes.TOP:
                        a = Math.max(0, (e.width - t.width) / 2), r = 0;
                        break;
                    case o.alignTypes.TOP_LEFT:
                        a = 0, r = 0;
                        break;
                    case o.alignTypes.TOP_RIGHT:
                        a = Math.max(0, e.width - t.width), r = 0;
                        break;
                    case o.alignTypes.BOTTOM:
                        a = Math.max(0, (e.width - t.width) / 2), r = Math.max(0, e.height - t.height);
                        break;
                    case o.alignTypes.BOTTOM_LEFT:
                        a = 0, r = Math.max(0, e.height - t.height);
                        break;
                    case o.alignTypes.BOTTOM_RIGHT:
                        a = Math.max(0, e.width - t.width), r = Math.max(0, e.height - t.height);
                        break;
                    case o.alignTypes.LEFT:
                        a = 0, r = Math.max(0, (e.height - t.height) / 2);
                        break;
                    case o.alignTypes.RIGHT:
                        a = Math.max(0, e.width - t.width), r = Math.max(0, (e.height - t.height) / 2)
                }
                return {
                    x: e.x ? e.x + a : a,
                    y: e.y ? e.y + r : r,
                    width: Math.min(e.width, t.width),
                    height: Math.min(e.height, t.height)
                }
            },
            getOverlappingRect: function(e, t) {
                var i = Math.max(0, Math.min(e.width, t.x + t.width) - Math.max(0, t.x)),
                    a = Math.max(0, Math.min(e.height, t.y + t.height) - Math.max(0, t.y));
                return i && a && (e.width !== i || e.height !== a) ? {
                    x: Math.max(0, t.x),
                    y: Math.max(0, t.y),
                    width: i,
                    height: a
                } : null
            },
            getScaleFactor: l,
            getTransformData: function(e, t, i, a, r, n) {
                var s = function(e, t, i, a, r) {
                        var n, s, c;
                        if (n = l(e, t, i, a, r), r === o.transformTypes.FILL ? (s = i, c = a) : r === o.transformTypes.FIT && (s = e * n, c = t * n), s * c > o.SAFE_TRANSFORMED_AREA) {
                            var h = Math.sqrt(o.SAFE_TRANSFORMED_AREA / (s * c));
                            n = l(e, t, s *= h, c *= h, r)
                        }
                        return {
                            scaleFactor: n,
                            width: s,
                            height: c
                        }
                    }(e = e || i.width, t = t || i.height, i.width * a, i.height * a, r),
                    c = s.scaleFactor;
                return function(e, t, i, a, r, n, s) {
                    var c, h, u = function(e, t, i, a) {
                            return {
                                classic: g,
                                auto: T,
                                super: f
                            }[a](e, t, i)
                        }(e, t, n, r),
                        p = u.optimizedScaleFactor,
                        l = u.upscaleMethodValue,
                        m = u.forceUSM;
                    if (n <= p) return {
                        width: i,
                        height: a,
                        scaleFactor: n,
                        upscaleMethodValue: l,
                        forceUSM: m,
                        cssUpscaleNeeded: !1
                    };
                    switch (s) {
                        case o.transformTypes.FILL:
                            c = i * (p / n), h = a * (p / n);
                            break;
                        case o.transformTypes.FIT:
                            c = e * p, h = t * p
                    }
                    return {
                        width: c,
                        height: h,
                        scaleFactor: p,
                        upscaleMethodValue: l,
                        forceUSM: m,
                        cssUpscaleNeeded: !0
                    }
                }(e, t, s.width, s.height, n, c, r)
            },
            getDevicePixelRatio: function(e) {
                return Math.min(e.pixelAspectRatio || 1, o.MAX_DEVICE_PIXEL_RATIO)
            },
            getAlignment: function(e) {
                return o.alignTypesMap[e.alignment] || o.alignTypesMap[o.alignTypes.CENTER]
            },
            getPreferredImageQuality: function(e, t) {
                return o.imageScaleDefaults[m(e, t)].quality
            },
            getDimension: function(e, t, i, a, r) {
                var n = l(e, t, i, a, r);
                return {
                    width: Math.round(e * n),
                    height: Math.round(t * n)
                }
            },
            getFocalPoint: function(e) {
                var t = null;
                return "number" != typeof e.x || isNaN(e.x) || "number" != typeof e.y || isNaN(e.y) || (t = {
                    x: _(Math.max(0, Math.min(100, e.x)) / 100, 2),
                    y: _(Math.max(0, Math.min(100, e.y)) / 100, 2)
                }), t
            },
            getUpscaleString: function(e) {
                return e && e.upscaleMethod && "string" == typeof e.upscaleMethod && o.upscaleMethods[e.upscaleMethod.toUpperCase()] || o.upscaleMethods.AUTO
            },
            roundToFixed: _
        }
    }, function(e, t, i) {
        "use strict";

        function a(e) {
            return (a = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(e) {
                return typeof e
            } : function(e) {
                return e && "function" == typeof Symbol && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e
            })(e)
        }
        e.exports = {
            assign: function() {
                for (var e = arguments[0] || {}, t = Array.prototype.slice.call(arguments, 1, arguments.length), i = 0; i < t.length; i++) {
                    var a = t[i];
                    for (var r in a) a.hasOwnProperty(r) && (e[r] = a[r])
                }
                return e
            },
            includes: function(e, t) {
                return e.indexOf ? e.indexOf(t) > -1 : !(!e || "object" !== a(e)) && Object.keys(e).some((function(i) {
                    return e[i] === t
                }))
            },
            last: function(e) {
                return e[e.length - 1]
            },
            template: function(e) {
                return function(t) {
                    var i = e;
                    for (var a in t) t.hasOwnProperty(a) && (i = i.replace(new RegExp("\\${" + a + "}", "g"), t[a]));
                    return i
                }
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a = {
            isWEBP: {
                lossless: !1,
                lossy: !1,
                alpha: !1,
                animation: !1
            },
            isObjectFitBrowser: !0
        };
        e.exports = {
            getFeature: function(e) {
                return a[e]
            },
            setFeature: function(e, t) {
                a[e] = t
            }
        }
    }, function(e, t, i) {
        "use strict";

        function a(e, t) {
            var i = Object.keys(e);
            if (Object.getOwnPropertySymbols) {
                var a = Object.getOwnPropertySymbols(e);
                t && (a = a.filter((function(t) {
                    return Object.getOwnPropertyDescriptor(e, t).enumerable
                }))), i.push.apply(i, a)
            }
            return i
        }

        function r(e, t, i) {
            return t in e ? Object.defineProperty(e, t, {
                value: i,
                enumerable: !0,
                configurable: !0,
                writable: !0
            }) : e[t] = i, e
        }
        var n = i(1),
            o = i(7),
            s = i(8),
            c = i(0);
        e.exports = {
            getTransform: function(e, t, i, a) {
                var r = n.isSEOBot(a),
                    c = n.getFileType(t.id),
                    h = n.getFileName(t.id, t.name),
                    u = n.getFileExtension(t.id),
                    p = !r && n.isWEBPBrowserSupport(c),
                    l = r ? 1 : n.getDevicePixelRatio(i),
                    g = {
                        fileName: h,
                        fileExtension: u,
                        fileType: c,
                        isWEBPSupport: p,
                        fittingType: e,
                        preferredExtension: p ? "webp" : u,
                        src: {
                            id: t.id,
                            width: t.width,
                            height: t.height,
                            isCropped: !1
                        },
                        focalPoint: {
                            x: t.focalPoint && t.focalPoint.x,
                            y: t.focalPoint && t.focalPoint.y
                        },
                        parts: [],
                        devicePixelRatio: l,
                        quality: 0,
                        upscaleMethod: n.getUpscaleString(a),
                        progressive: !0,
                        watermark: "",
                        unsharpMask: {},
                        filters: {}
                    };
                return n.isImageTransformApplicable(t.id) && (o.setTransformParts(g, t, i), s.setTransformOptions(g, a)), g
            },
            getTarget: function(e, t, i) {
                var o = function(e) {
                        for (var t = 1; t < arguments.length; t++) {
                            var i = null != arguments[t] ? arguments[t] : {};
                            t % 2 ? a(Object(i), !0).forEach((function(t) {
                                r(e, t, i[t])
                            })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(i)) : a(Object(i)).forEach((function(t) {
                                Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(i, t))
                            }))
                        }
                        return e
                    }({}, i),
                    s = n.isMobile();
                switch (e) {
                    case c.fittingTypes.LEGACY_BG_FIT_AND_TILE:
                    case c.fittingTypes.LEGACY_BG_FIT_AND_TILE_HORIZONTAL:
                    case c.fittingTypes.LEGACY_BG_FIT_AND_TILE_VERTICAL:
                    case c.fittingTypes.LEGACY_BG_NORMAL:
                        var h = s ? c.MOBILE_MAX_BG_SITE_LEGACY_WIDTH : c.DSKTP_MAX_BG_SITE_LEGACY_WIDTH,
                            u = s ? c.MOBILE_MAX_BG_SITE_LEGACY_HEIGHT : c.DSKTP_MAX_BG_SITE_LEGACY_HEIGHT;
                        o.width = Math.min(h, t.width), o.height = Math.min(u, Math.round(o.width / (t.width / t.height))), o.pixelAspectRatio = 1
                }
                return o
            }
        }
    }, function(e, t, i) {
        e.exports = function() {
            "use strict";
            /*!
             * Based on Zepto's detect module - https://github.com/madrobby/zepto/blob/master/src/detect.js#files
             * Zepto.js may be freely distributed under the MIT license. See: https://github.com/madrobby/zepto/blob/master/MIT-LICENSE
             *
             * note - MS Edge detection was added here, which Zepto does not have.
             */
            return function(e) {
                var t = {},
                    i = {};
                if (!e) return {
                    browser: i,
                    os: t
                };
                var a = e.match(/Web[kK]it[\/]{0,1}([\d.]+)/),
                    r = e.match(/(Android);?[\s\/]+([\d.]+)?/),
                    n = !!e.match(/\(Macintosh\; Intel /),
                    o = e.match(/(iPad).*OS\s([\d_]+)/),
                    s = e.match(/(iPod)(.*OS\s([\d_]+))?/),
                    c = !o && e.match(/(iPhone\sOS)\s([\d_]+)/),
                    h = e.match(/(webOS|hpwOS)[\s\/]([\d.]+)/),
                    u = e.match(/Windows Phone ([\d.]+)/),
                    p = h && e.match(/TouchPad/),
                    l = e.match(/Kindle\/([\d.]+)/),
                    g = e.match(/Silk\/([\d._]+)/),
                    T = e.match(/(BlackBerry).*Version\/([\d.]+)/),
                    f = e.match(/(BB10).*Version\/([\d.]+)/),
                    m = e.match(/(RIM\sTablet\sOS)\s([\d.]+)/),
                    _ = e.match(/PlayBook/),
                    d = e.match(/Chrome\/([\d.]+)/) || e.match(/CriOS\/([\d.]+)/),
                    A = e.match(/Firefox\/([\d.]+)/),
                    I = e.match(/MSIE\s([\d.]+)/) || e.match(/Trident\/[\d](?=[^\?]+).*rv:([0-9.].)/),
                    E = !d && e.match(/(iPhone|iPod|iPad).*AppleWebKit(?!.*Safari)/),
                    L = E || e.match(/Version\/([\d.]+)([^S](Safari)|[^M]*(Mobile)[^S]*(Safari))/),
                    b = e.match(/Edge\/(\d{2,}\.[\d\w]+)$/),
                    y = e.match(/Opera Mini/);
                return i.webkit = a && !b, i.webkit && (i.version = a[1]), r && (t.android = !0, t.version = r[2]), c && !s && (t.ios = t.iphone = !0, t.version = c[2].replace(/_/g, ".")), o && (t.ios = t.ipad = !0, t.version = o[2].replace(/_/g, ".")), s && (t.ios = t.ipod = !0, t.version = s[3] ? s[3].replace(/_/g, ".") : null), u && (t.wp = !0, t.version = u[1]), h && (t.webos = !0, t.version = h[2]), p && (t.touchpad = !0), T && (t.blackberry = !0, t.version = T[2]), f && (t.bb10 = !0, t.version = f[2]), m && (t.rimtabletos = !0, t.version = m[2]), _ && (i.playbook = !0), l && (t.kindle = !0, t.version = l[1]), g && (i.silk = !0, i.version = g[1]), !g && t.android && e.match(/Kindle Fire/) && (i.silk = !0), d && !b && (i.chrome = !0, i.version = d[1]), A && !b && (i.firefox = !0, i.version = A[1]), I && (i.ie = !0, i.version = I[1]), L && (n || t.ios) && (i.safari = !0, n && (i.version = L[1])), E && (i.webview = !0), b && (i.edge = !0, i.version = b[1]), y && (i.operaMini = !0), t.tablet = !!(o || _ || r && !e.match(/Mobile/) || A && e.match(/Tablet/) || (I || b) && !e.match(/Phone/) && e.match(/Touch/)), t.phone = !(t.tablet || t.ipod || !(r || c || h || T || f || d && e.match(/Android/) || d && e.match(/CriOS\/([\d.]+)/) || A && e.match(/Mobile/) || I && e.match(/Touch/))), t.mac = n, t.googleBot = !!e.match(/Googlebot\/2.1/), {
                    browser: i,
                    os: t
                }
            }
        }()
    }, function(e, t, i) {
        "use strict";
        var a = i(3),
            r = i(0);

        function n(e) {
            var t = a.getFeature("isWEBP"),
                i = new window.Image;
            i.onload = function() {
                t[e] = i.width > 0 && i.height > 0, a.setFeature("isWEBP", t)
            }, i.onerror = function() {
                t[e] = !1, a.setFeature("isWEBP", t)
            }, i.src = "data:image/webp;base64," + {
                lossy: "UklGRiIAAABXRUJQVlA4IBYAAAAwAQCdASoBAAEADsD+JaQAA3AAAAAA",
                lossless: "UklGRhoAAABXRUJQVlA4TA0AAAAvAAAAEAcQERGIiP4HAA==",
                alpha: "UklGRkoAAABXRUJQVlA4WAoAAAAQAAAAAAAAAAAAQUxQSAwAAAARBxAR/Q9ERP8DAABWUDggGAAAABQBAJ0BKgEAAQAAAP4AAA3AAP7mtQAAAA==",
                animation: "UklGRlIAAABXRUJQVlA4WAoAAAASAAAAAAAAAAAAQU5JTQYAAAD/////AABBTk1GJgAAAAAAAAAAAAAAAAAAAGQAAABWUDhMDQAAAC8AAAAQBxAREYiI/gcA"
            }[e]
        }
        e.exports = {
            checkSupportByUserAgent: function(e, t) {
                var i = t.browser,
                    n = t.os;
                if (e) {
                    var o, s = parseFloat(i.version),
                        c = parseFloat(n.version),
                        h = new RegExp(/AppleWebKit\/([\d.]+)/),
                        u = null === h.exec(e) ? null : parseFloat(h.exec(e)[1]),
                        p = function(e, t, i, a, r) {
                            var n = !e.phone && !e.tablet && i.chrome && a >= 23,
                                o = e.android && (e.phone || e.tablet) && i.webkit && i.chrome && a >= 25,
                                s = e.android && r < 535 && (e.phone || e.tablet) && t >= 4.2 && i.webkit && !i.safari,
                                c = i.edge && a >= 18,
                                h = !e.firefoxos && i.firefox && !i.webkit && a >= 65;
                            return !!(n || o || s || c || h)
                        }(n, c, i, s, u);
                    a.setFeature("isWEBP", ((o = {})[r.webp.LOSSY] = function(e, t, i, a, r) {
                        var n = !e.phone && !e.tablet && i.chrome && a >= 17,
                            o = e.android && (e.phone || e.tablet) && i.webkit && i.chrome && a >= 25,
                            s = e.android && r < 535 && (e.phone || e.tablet) && t >= 4 && i.webkit,
                            c = i.edge && a >= 18,
                            h = !e.firefoxos && i.firefox && !i.webkit && a >= 65;
                        return !!(n || o || s || c || h)
                    }(n, c, i, s, u), o[r.webp.LOSSLESS] = p, o[r.webp.ALPHA] = p, o[r.webp.ANIMATION] = function(e, t, i, a) {
                        var r = !e.ios && i.chrome && a >= 32,
                            n = i.edge && a >= 18,
                            o = !e.firefoxos && i.firefox && !i.webkit && a >= 65;
                        return !!(r || n || o)
                    }(n, 0, i, s), o))
                }
            },
            checkSupportByFeatureDetection: function() {
                n(r.webp.LOSSY), n(r.webp.LOSSLESS), n(r.webp.ALPHA), n(r.webp.ANIMATION)
            },
            isWEBPBrowserSupport: function(e) {
                var t = a.getFeature("isWEBP"),
                    i = e === r.fileType.JPG && t[r.webp.LOSSY],
                    n = e === r.fileType.PNG && t[r.webp.LOSSLESS],
                    o = e === r.fileType.PNG && t[r.webp.ALPHA];
                return i || n && o
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(2),
            r = i(0),
            n = i(1);

        function o(e, t) {
            var i = n.getTransformData(e.src.width, e.src.height, t, e.devicePixelRatio, r.transformTypes.FIT, e.upscaleMethod);
            return {
                transformType: r.transformTypes.FILL,
                width: Math.round(i.width),
                height: Math.round(i.height),
                alignment: r.alignTypesMap.center,
                upscale: i.scaleFactor > 1,
                forceUSM: i.forceUSM,
                scaleFactor: i.scaleFactor,
                cssUpscaleNeeded: i.cssUpscaleNeeded,
                upscaleMethodValue: i.upscaleMethodValue
            }
        }

        function s(e) {
            return {
                transformType: r.transformTypes.CROP,
                x: Math.round(e.x),
                y: Math.round(e.y),
                width: Math.round(e.width),
                height: Math.round(e.height),
                upscale: !1,
                forceUSM: !1,
                scaleFactor: 1,
                cssUpscaleNeeded: !1
            }
        }
        e.exports = {
            setTransformParts: function(e, t, i) {
                var c;
                switch (t.crop && (c = n.getOverlappingRect(t, t.crop)) && (e.src.width = c.width, e.src.height = c.height, e.src.cropped = !0, e.parts.push(s(c))), e.fittingType) {
                    case r.fittingTypes.SCALE_TO_FIT:
                    case r.fittingTypes.LEGACY_FIT_WIDTH:
                    case r.fittingTypes.LEGACY_FIT_HEIGHT:
                    case r.fittingTypes.LEGACY_FULL:
                    case r.fittingTypes.FIT_AND_TILE:
                    case r.fittingTypes.LEGACY_BG_FIT_AND_TILE:
                    case r.fittingTypes.LEGACY_BG_FIT_AND_TILE_HORIZONTAL:
                    case r.fittingTypes.LEGACY_BG_FIT_AND_TILE_VERTICAL:
                    case r.fittingTypes.LEGACY_BG_NORMAL:
                        e.parts.push(o(e, i));
                        break;
                    case r.fittingTypes.SCALE_TO_FILL:
                        e.parts.push(function(e, t) {
                            var i = n.getTransformData(e.src.width, e.src.height, t, e.devicePixelRatio, r.transformTypes.FILL, e.upscaleMethod),
                                a = n.getFocalPoint(e.focalPoint);
                            return {
                                transformType: a ? r.transformTypes.FILL_FOCAL : r.transformTypes.FILL,
                                width: Math.round(i.width),
                                height: Math.round(i.height),
                                alignment: n.getAlignment(t),
                                focalPointX: a && a.x,
                                focalPointY: a && a.y,
                                upscale: i.scaleFactor > 1,
                                forceUSM: i.forceUSM,
                                scaleFactor: i.scaleFactor,
                                cssUpscaleNeeded: i.cssUpscaleNeeded,
                                upscaleMethodValue: i.upscaleMethodValue
                            }
                        }(e, i));
                        break;
                    case r.fittingTypes.STRETCH:
                        e.parts.push(function(e, t) {
                            var i = n.getScaleFactor(e.src.width, e.src.height, t.width, t.height, r.transformTypes.FILL),
                                s = a.assign({}, t);
                            return s.width = e.src.width * i, s.height = e.src.height * i, o(e, s)
                        }(e, i));
                        break;
                    case r.fittingTypes.TILE_HORIZONTAL:
                    case r.fittingTypes.TILE_VERTICAL:
                    case r.fittingTypes.TILE:
                    case r.fittingTypes.LEGACY_ORIGINAL_SIZE:
                    case r.fittingTypes.ORIGINAL_SIZE:
                        c = n.getAlignedRect(e.src, i, i.alignment), e.src.isCropped ? (a.assign(e.parts[0], c), e.src.width = c.width, e.src.height = c.height) : e.parts.push(s(c));
                        break;
                    case r.fittingTypes.LEGACY_STRIP_TILE_HORIZONTAL:
                    case r.fittingTypes.LEGACY_STRIP_TILE_VERTICAL:
                    case r.fittingTypes.LEGACY_STRIP_TILE:
                    case r.fittingTypes.LEGACY_STRIP_ORIGINAL_SIZE:
                        e.parts.push(function(e) {
                            return {
                                transformType: r.transformTypes.LEGACY_CROP,
                                width: Math.round(e.width),
                                height: Math.round(e.height),
                                alignment: n.getAlignment(e),
                                upscale: !1,
                                forceUSM: !1,
                                scaleFactor: 1,
                                cssUpscaleNeeded: !1
                            }
                        }(i));
                        break;
                    case r.fittingTypes.LEGACY_STRIP_SCALE_TO_FIT:
                    case r.fittingTypes.LEGACY_STRIP_FIT_AND_TILE:
                        e.parts.push(function(e) {
                            return {
                                transformType: r.transformTypes.FIT,
                                width: Math.round(e.width),
                                height: Math.round(e.height),
                                upscale: !1,
                                forceUSM: !0,
                                scaleFactor: 1,
                                cssUpscaleNeeded: !1
                            }
                        }(i));
                        break;
                    case r.fittingTypes.LEGACY_STRIP_SCALE_TO_FILL:
                        e.parts.push(function(e) {
                            return {
                                transformType: r.transformTypes.LEGACY_FILL,
                                width: Math.round(e.width),
                                height: Math.round(e.height),
                                alignment: n.getAlignment(e),
                                upscale: !1,
                                forceUSM: !0,
                                scaleFactor: 1,
                                cssUpscaleNeeded: !1
                            }
                        }(i))
                }
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(2),
            r = i(0),
            n = i(1);

        function o(e, t, i) {
            return !isNaN(e) && "number" == typeof e && 0 !== e && e >= t && e <= i
        }
        e.exports = {
            setTransformOptions: function(e, t) {
                t = t || {}, e.quality = function(e, t) {
                    var i = e.fileType === r.fileType.PNG && e.isWEBPSupport;
                    if (e.fileType === r.fileType.JPG || i) {
                        var o = a.last(e.parts),
                            s = n.getPreferredImageQuality(o.width, o.height),
                            c = t.quality && t.quality >= 5 && t.quality <= 90 ? t.quality : s;
                        return c = i ? c + 5 : c, parseInt(c, 10)
                    }
                    return 0
                }(e, t), e.progressive = function(e) {
                    return !1 !== e.progressive
                }(t), e.watermark = function(e) {
                    return e.watermark
                }(t), e.unsharpMask = function(e, t) {
                    var i;
                    ! function(e) {
                        e = e || {};
                        var t = !isNaN(e.radius) && "number" == typeof e.radius && e.radius >= .1 && e.radius <= 500,
                            i = !isNaN(e.amount) && "number" == typeof e.amount && e.amount >= 0 && e.amount <= 10,
                            a = !isNaN(e.threshold) && "number" == typeof e.threshold && e.threshold >= 0 && e.threshold <= 255;
                        return t && i && a
                    }(t.unsharpMask) ? function(e) {
                        return e = e || {}, !isNaN(e.radius) && "number" == typeof e.radius && 0 === e.radius && !isNaN(e.amount) && "number" == typeof e.amount && 0 === e.amount && !isNaN(e.threshold) && "number" == typeof e.threshold && 0 === e.threshold
                    }(t.unsharpMask) || function(e) {
                        var t = a.last(e.parts);
                        return !(t.scaleFactor >= 1) || t.forceUSM
                    }(e) && (i = r.defaultUSM) : i = {
                        radius: t.unsharpMask.radius,
                        amount: t.unsharpMask.amount,
                        threshold: t.unsharpMask.threshold
                    };
                    i && (i.radius = n.roundToFixed(i.radius, 2), i.amount = n.roundToFixed(i.amount, 2), i.threshold = n.roundToFixed(i.threshold, 2));
                    return i
                }(e, t), e.filters = function(e) {
                    var t = e.filters || {},
                        i = {};
                    o(t[r.imageFilters.CONTRAST], -100, 100) && (i[r.imageFilters.CONTRAST] = t[r.imageFilters.CONTRAST]);
                    o(t[r.imageFilters.BRIGHTNESS], -100, 100) && (i[r.imageFilters.BRIGHTNESS] = t[r.imageFilters.BRIGHTNESS]);
                    o(t[r.imageFilters.SATURATION], -100, 100) && (i[r.imageFilters.SATURATION] = t[r.imageFilters.SATURATION]);
                    o(t[r.imageFilters.HUE], -180, 180) && (i[r.imageFilters.HUE] = t[r.imageFilters.HUE]);
                    o(t[r.imageFilters.BLUR], 0, 100) && (i[r.imageFilters.BLUR] = t[r.imageFilters.BLUR]);
                    return i
                }(t)
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(0),
            r = i(1),
            n = i(10),
            o = i(4);
        e.exports = function(e, t, i, s, c) {
            var h = a.emptyData.uri;
            return r.isImageTransformApplicable(t.id) ? (c = c || o.getTransform(e, t, i, s, c), h = n.getImageURI(c)) : h = t.id, h
        }
    }, function(e, t, i) {
        "use strict";
        var a, r = i(2),
            n = i(0),
            o = r.template("fit/w_${width},h_${height}"),
            s = r.template("fill/w_${width},h_${height},al_${alignment}"),
            c = r.template("fill/w_${width},h_${height},fp_${focalPointX}_${focalPointY}"),
            h = r.template("crop/x_${x},y_${y},w_${width},h_${height}"),
            u = r.template("crop/w_${width},h_${height},al_${alignment}"),
            p = r.template("fill/w_${width},h_${height},al_${alignment}"),
            l = r.template(",lg_${upscaleMethodValue}"),
            g = r.template(",q_${quality}"),
            T = r.template(",usm_${radius}_${amount}_${threshold}"),
            f = r.template(",bl"),
            m = r.template(",wm_${watermark}"),
            _ = ((a = {})[n.imageFilters.CONTRAST] = r.template(",con_${contrast}"), a[n.imageFilters.BRIGHTNESS] = r.template(",br_${brightness}"), a[n.imageFilters.SATURATION] = r.template(",sat_${saturation}"), a[n.imageFilters.HUE] = r.template(",hue_${hue}"), a[n.imageFilters.BLUR] = r.template(",blur_${blur}"), a);
        e.exports = {
            getImageURI: function(e) {
                var t = [];
                e.parts.forEach((function(e) {
                    switch (e.transformType) {
                        case n.transformTypes.CROP:
                            t.push(h(e));
                            break;
                        case n.transformTypes.LEGACY_CROP:
                            t.push(u(e));
                            break;
                        case n.transformTypes.LEGACY_FILL:
                            var i = p(e);
                            e.upscale && (i += l(e)), t.push(i);
                            break;
                        case n.transformTypes.FIT:
                            var a = o(e);
                            e.upscale && (a += l(e)), t.push(a);
                            break;
                        case n.transformTypes.FILL:
                            var r = s(e);
                            e.upscale && (r += l(e)), t.push(r);
                            break;
                        case n.transformTypes.FILL_FOCAL:
                            var g = c(e);
                            e.upscale && (g += l(e)), t.push(g)
                    }
                }));
                var i = t.join("/");
                return e.quality && (i += g(e)), e.unsharpMask && (i += T(e.unsharpMask)), e.progressive || (i += f(e)), e.watermark && (i += m(e)), e.filters && (i += Object.keys(e.filters).map((function(t) {
                    return _[t](e.filters)
                })).join("")), e.src.id + "/" + n.API_VERSION + "/" + i + "/" + e.fileName + "." + e.preferredExtension
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(2),
            r = i(0),
            n = i(1),
            o = i(12),
            s = i(4),
            c = i(9);
        n.populateGlobalFeatureSupport(), e.exports = {
            populateGlobalFeatureSupport: n.populateGlobalFeatureSupport,
            getData: function(e, t, i, h) {
                var u = {};
                if (n.isValidRequest(e, t, i)) {
                    var p = s.getTarget(e, t, i),
                        l = s.getTransform(e, t, p, h);
                    u.uri = c(e, t, p, h, l), a.assign(u, o.getAttributes(l, p))
                } else u = r.emptyData;
                return u
            },
            fittingTypes: r.fittingTypes,
            alignTypes: r.alignTypes,
            htmlTag: r.htmlTag,
            upscaleMethods: r.upscaleMethods
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(0),
            r = i(1),
            n = i(13),
            o = i(14),
            s = i(15),
            c = i(16);
        e.exports = {
            getAttributes: function(e, t) {
                return (t.htmlTag === a.htmlTag.BG ? n : t.htmlTag === a.htmlTag.SVG ? s : r.isObjectFitBrowserSupport() ? o : c).get(e, t)
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(0);
        e.exports = {
            get: function(e, t) {
                var i = {
                        css: {
                            container: {}
                        }
                    },
                    r = i.css,
                    n = a.alignTypes,
                    o = e.fittingType,
                    s = a.fittingTypes;
                switch (o) {
                    case s.ORIGINAL_SIZE:
                    case s.LEGACY_ORIGINAL_SIZE:
                    case s.LEGACY_STRIP_ORIGINAL_SIZE:
                        r.container.backgroundSize = "auto", r.container.backgroundRepeat = "no-repeat";
                        break;
                    case s.SCALE_TO_FIT:
                    case s.LEGACY_STRIP_SCALE_TO_FIT:
                        r.container.backgroundSize = "contain", r.container.backgroundRepeat = "no-repeat";
                        break;
                    case s.STRETCH:
                        r.container.backgroundSize = "100% 100%", r.container.backgroundRepeat = "no-repeat";
                        break;
                    case s.SCALE_TO_FILL:
                    case s.LEGACY_STRIP_SCALE_TO_FILL:
                        r.container.backgroundSize = "cover", r.container.backgroundRepeat = "no-repeat";
                        break;
                    case s.TILE_HORIZONTAL:
                    case s.LEGACY_STRIP_TILE_HORIZONTAL:
                        r.container.backgroundSize = "auto", r.container.backgroundRepeat = "repeat-x";
                        break;
                    case s.TILE_VERTICAL:
                    case s.LEGACY_STRIP_TILE_VERTICAL:
                        r.container.backgroundSize = "auto", r.container.backgroundRepeat = "repeat-y";
                        break;
                    case s.TILE:
                    case s.LEGACY_STRIP_TILE:
                        r.container.backgroundSize = "auto", r.container.backgroundRepeat = "repeat";
                        break;
                    case s.FIT_AND_TILE:
                    case s.LEGACY_STRIP_FIT_AND_TILE:
                        r.container.backgroundSize = "contain", r.container.backgroundRepeat = "repeat";
                        break;
                    case s.LEGACY_BG_FIT_AND_TILE:
                        r.container.backgroundSize = "auto", r.container.backgroundRepeat = "repeat";
                        break;
                    case s.LEGACY_BG_FIT_AND_TILE_HORIZONTAL:
                        r.container.backgroundSize = "auto", r.container.backgroundRepeat = "repeat-x";
                        break;
                    case s.LEGACY_BG_FIT_AND_TILE_VERTICAL:
                        r.container.backgroundSize = "auto", r.container.backgroundRepeat = "repeat-y";
                        break;
                    case s.LEGACY_BG_NORMAL:
                        r.container.backgroundSize = "auto", r.container.backgroundRepeat = "no-repeat"
                }
                switch (t.alignment) {
                    case n.CENTER:
                        r.container.backgroundPosition = "center center";
                        break;
                    case n.LEFT:
                        r.container.backgroundPosition = "left center";
                        break;
                    case n.RIGHT:
                        r.container.backgroundPosition = "right center";
                        break;
                    case n.TOP:
                        r.container.backgroundPosition = "center top";
                        break;
                    case n.BOTTOM:
                        r.container.backgroundPosition = "center bottom";
                        break;
                    case n.TOP_RIGHT:
                        r.container.backgroundPosition = "right top";
                        break;
                    case n.TOP_LEFT:
                        r.container.backgroundPosition = "left top";
                        break;
                    case n.BOTTOM_RIGHT:
                        r.container.backgroundPosition = "right bottom";
                        break;
                    case n.BOTTOM_LEFT:
                        r.container.backgroundPosition = "left bottom"
                }
                return i
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a, r = i(0),
            n = r.alignTypes;
        var o = ((a = {})[n.CENTER] = "center", a[n.TOP] = "top", a[n.TOP_LEFT] = "top left", a[n.TOP_RIGHT] = "top right", a[n.BOTTOM] = "bottom", a[n.BOTTOM_LEFT] = "bottom left", a[n.BOTTOM_RIGHT] = "bottom right", a[n.LEFT] = "left", a[n.RIGHT] = "right", a),
            s = {
                position: "absolute",
                top: "auto",
                right: "auto",
                bottom: "auto",
                left: "auto"
            };
        e.exports = {
            get: function(e, t) {
                var i = {
                        css: {
                            container: {},
                            img: {}
                        }
                    },
                    a = i.css,
                    c = e.fittingType,
                    h = t.alignment,
                    u = r.fittingTypes;
                switch (a.container.position = "relative", c) {
                    case u.ORIGINAL_SIZE:
                    case u.LEGACY_ORIGINAL_SIZE:
                        e.parts && e.parts.length ? (a.img.width = e.parts[0].width, a.img.height = e.parts[0].height) : (a.img.width = e.src.width, a.img.height = e.src.height);
                        break;
                    case u.SCALE_TO_FIT:
                    case u.LEGACY_FIT_WIDTH:
                    case u.LEGACY_FIT_HEIGHT:
                    case u.LEGACY_FULL:
                        a.img.width = t.width, a.img.height = t.height, a.img.objectFit = "contain", a.img.objectPosition = o[h] || "unset";
                        break;
                    case u.LEGACY_BG_NORMAL:
                        a.img.width = "100%", a.img.height = "100%", a.img.objectFit = "none", a.img.objectPosition = o[h] || "unset";
                        break;
                    case u.STRETCH:
                        a.img.width = t.width, a.img.height = t.height, a.img.objectFit = "fill";
                        break;
                    case u.SCALE_TO_FILL:
                        a.img.width = t.width, a.img.height = t.height, a.img.objectFit = "cover"
                }
                if ("number" == typeof a.img.width && (a.img.width !== t.width || a.img.height !== t.height)) {
                    var p = Math.round((t.height - a.img.height) / 2),
                        l = Math.round((t.width - a.img.width) / 2);
                    Object.assign(a.img, s, function(e, t, i) {
                        var a;
                        return (a = {})[n.TOP_LEFT] = {
                            top: 0,
                            left: 0
                        }, a[n.TOP_RIGHT] = {
                            top: 0,
                            right: 0
                        }, a[n.TOP] = {
                            top: 0,
                            left: t
                        }, a[n.BOTTOM_LEFT] = {
                            top: 0,
                            left: 0
                        }, a[n.BOTTOM_RIGHT] = {
                            bottom: 0,
                            right: 0
                        }, a[n.BOTTOM] = {
                            bottom: 0,
                            left: t
                        }, a[n.RIGHT] = {
                            top: e,
                            right: 0
                        }, a[n.LEFT] = {
                            top: e,
                            left: 0
                        }, a[n.CENTER] = {
                            width: i.width,
                            height: i.height,
                            objectFit: "none"
                        }, a
                    }(p, l, t)[h])
                }
                return i
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(0),
            r = i(1);
        e.exports = {
            get: function(e, t) {
                var i, n = {
                        css: {
                            container: {}
                        },
                        attr: {
                            container: {},
                            img: {}
                        }
                    },
                    o = n.css,
                    s = n.attr,
                    c = e.fittingType,
                    h = a.fittingTypes,
                    u = t.alignment,
                    p = a.alignTypes,
                    l = e.src.width,
                    g = e.src.height;
                switch (o.container.position = "relative", c) {
                    case h.ORIGINAL_SIZE:
                    case h.LEGACY_ORIGINAL_SIZE:
                    case h.TILE:
                        e.parts && e.parts.length ? (s.img.width = e.parts[0].width, s.img.height = e.parts[0].height) : (s.img.width = l, s.img.height = g), s.img.preserveAspectRatio = "xMidYMid slice";
                        break;
                    case h.SCALE_TO_FIT:
                    case h.LEGACY_FIT_WIDTH:
                    case h.LEGACY_FIT_HEIGHT:
                    case h.LEGACY_FULL:
                        s.img.width = "100%", s.img.height = "100%", s.img.transform = "", s.img.preserveAspectRatio = "";
                        break;
                    case h.STRETCH:
                        s.img.width = t.width, s.img.height = t.height, s.img.x = 0, s.img.y = 0, s.img.transform = "", s.img.preserveAspectRatio = "none";
                        break;
                    case h.SCALE_TO_FILL:
                        r.isImageTransformApplicable(e.src.id) ? (s.img.width = t.width, s.img.height = t.height) : (i = r.getDimension(l, g, t.width, t.height, a.transformTypes.FILL), s.img.width = i.width, s.img.height = i.height), s.img.x = 0, s.img.y = 0, s.img.transform = "", s.img.preserveAspectRatio = "xMidYMid slice"
                }
                if (s.img.width !== t.width || s.img.height !== t.height) {
                    var T, f, m = 0,
                        _ = 0;
                    c === h.TILE ? (T = t.width % s.img.width, f = t.height % s.img.height) : (T = t.width - s.img.width, f = t.height - s.img.height);
                    var d = Math.round(T / 2),
                        A = Math.round(f / 2);
                    switch (u) {
                        case p.TOP_LEFT:
                            m = 0, _ = 0;
                            break;
                        case p.TOP:
                            m = d, _ = 0;
                            break;
                        case p.TOP_RIGHT:
                            m = T, _ = 0;
                            break;
                        case p.LEFT:
                            m = 0, _ = A;
                            break;
                        case p.CENTER:
                            m = d, _ = A;
                            break;
                        case p.RIGHT:
                            m = T, _ = A;
                            break;
                        case p.BOTTOM_LEFT:
                            m = 0, _ = f;
                            break;
                        case p.BOTTOM:
                            m = d, _ = f;
                            break;
                        case p.BOTTOM_RIGHT:
                            m = T, _ = f
                    }
                    s.img.x = m, s.img.y = _
                }
                return s.container.width = t.width, s.container.height = t.height, s.container.viewBox = [0, 0, t.width, t.height].join(" "), n
            }
        }
    }, function(e, t, i) {
        "use strict";
        var a = i(2),
            r = i(0),
            n = i(1);
        e.exports = {
            get: function(e, t) {
                var i, o, s, c = {
                        css: {
                            container: {},
                            img: {}
                        }
                    },
                    h = c.css,
                    u = e.fittingType,
                    p = r.fittingTypes,
                    l = t.alignment,
                    g = r.alignTypes;
                switch (e.parts && e.parts.length ? (i = e.parts[0].width, o = e.parts[0].height) : (i = e.src.width, o = e.src.height), h.img.display = "block", h.container.position = "relative", h.img.position = "absolute", h.img.top = "auto", h.img.right = "auto", h.img.bottom = "auto", h.img.left = "auto", u) {
                    case p.ORIGINAL_SIZE:
                    case p.LEGACY_ORIGINAL_SIZE:
                        h.img.width = i, h.img.height = o;
                        break;
                    case p.SCALE_TO_FIT:
                    case p.LEGACY_FIT_WIDTH:
                    case p.LEGACY_FIT_HEIGHT:
                    case p.LEGACY_FULL:
                        s = n.getDimension(i, o, t.width, t.height, r.transformTypes.FIT), a.assign(h.img, s);
                        break;
                    case p.STRETCH:
                        h.img.width = t.width, h.img.height = t.height;
                        break;
                    case p.SCALE_TO_FILL:
                        var T = e.src.id;
                        n.isImageTypeSupported(T) || n.isExternalUrl(T) ? (h.img.width = t.width, h.img.height = t.height) : (s = n.getDimension(i, o, t.width, t.height, r.transformTypes.FILL), a.assign(h.img, s), h.container.overflow = "hidden")
                }
                if (h.img.width !== t.width || h.img.height !== t.height) {
                    var f = Math.round((t.height - h.img.height) / 2),
                        m = Math.round((t.width - h.img.width) / 2);
                    switch (l) {
                        default:
                            case g.CENTER:
                            h.img.top = f,
                        h.img.left = m;
                        break;
                        case g.LEFT:
                                h.img.left = 0,
                            h.img.top = f;
                            break;
                        case g.RIGHT:
                                h.img.right = 0,
                            h.img.top = f;
                            break;
                        case g.TOP:
                                h.img.left = m,
                            h.img.top = 0;
                            break;
                        case g.BOTTOM:
                                h.img.left = m,
                            h.img.bottom = 0;
                            break;
                        case g.TOP_RIGHT:
                                h.img.right = 0,
                            h.img.top = 0;
                            break;
                        case g.TOP_LEFT:
                                h.img.left = 0,
                            h.img.top = 0;
                            break;
                        case g.BOTTOM_RIGHT:
                                h.img.right = 0,
                            h.img.bottom = 0;
                            break;
                        case g.BOTTOM_LEFT:
                                h.img.left = 0,
                            h.img.bottom = 0
                    }
                }
                return c
            }
        }
    }])
}));
