// @wix/editor-elements-corvid-utils/dist/props-factories/hiddenCollapsedSDKFactory.js
var __assign = (this && this.__assign) || function() {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function(thisArg, _arguments, P, generator) {
    function adopt(value) {
        return value instanceof P ? value : new P(function(resolve) {
            resolve(value);
        });
    }
    return new(P || (P = Promise))(function(resolve, reject) {
        function fulfilled(value) {
            try {
                step(generator.next(value));
            } catch (e) {
                reject(e);
            }
        }

        function rejected(value) {
            try {
                step(generator["throw"](value));
            } catch (e) {
                reject(e);
            }
        }

        function step(result) {
            result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
        }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function(thisArg, body) {
    var _ = {
            label: 0,
            sent: function() {
                if (t[0] & 1) throw t[1];
                return t[1];
            },
            trys: [],
            ops: []
        },
        f, y, t, g;
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;

    function verb(n) {
        return function(v) {
            return step([n, v]);
        };
    }

    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [0];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [6, e];
            y = 0;
        } finally {
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
};
import {
    withValidation
} from '..';
import {
    createEffectValidation
} from './validations';
import {
    effectDefaultOptions,
    EFFECTS,
} from './animations';
export var createHiddenCollapsedSDKFactory = function(viewportHandlers) {
    var _hiddenCollapsedSDKFactory = function(_a) {
        var setStyles = _a.setStyles,
            metaData = _a.metaData,
            getSdkInstance = _a.getSdkInstance,
            runAnimation = _a.runAnimation;
        var validateEffect = createEffectValidation({
            compName: metaData.role,
        });
        var _state = {
            hidden: metaData.hiddenOnLoad,
            collapsed: metaData.collapsedOnLoad,
        };
        return {
            hide: function(effectName, effectOptions) {
                return __awaiter(void 0, void 0, void 0, function() {
                    var _a;
                    return __generator(this, function(_b) {
                        switch (_b.label) {
                            case 0:
                                if (_state.collapsed || _state.hidden) {
                                    _state.hidden = true;
                                    return [2 /*return*/ , Promise.resolve()];
                                }
                                if (!validateEffect({
                                        effectName: effectName,
                                        effectOptions: effectOptions,
                                        propertyName: 'hide',
                                    })) return [3 /*break*/ , 2];
                                return [4 /*yield*/ , runAnimation({
                                    animationDirection: EFFECTS.HIDE.suffix,
                                    effectName: effectName,
                                    effectOptions: __assign(__assign({}, effectDefaultOptions === null || effectDefaultOptions === void 0 ? void 0 : effectDefaultOptions[effectName]), effectOptions),
                                })];
                            case 1:
                                _b.sent();
                                _b.label = 2;
                            case 2:
                                setStyles({
                                    visibility: 'hidden !important'
                                });
                                _state.hidden = true;
                                (_a = viewportHandlers === null || viewportHandlers === void 0 ? void 0 : viewportHandlers.onViewPortLeave) === null || _a === void 0 ? void 0 : _a.forEach(function(cb) {
                                    return cb();
                                });
                                return [2 /*return*/ ];
                        }
                    });
                });
            },
            show: function(effectName, effectOptions) {
                return __awaiter(void 0, void 0, void 0, function() {
                    var _a;
                    return __generator(this, function(_b) {
                        switch (_b.label) {
                            case 0:
                                if (_state.collapsed || !_state.hidden) {
                                    _state.hidden = false;
                                    return [2 /*return*/ ];
                                }
                                if (!validateEffect({
                                        effectName: effectName,
                                        effectOptions: effectOptions,
                                        propertyName: 'show',
                                    })) return [3 /*break*/ , 2];
                                return [4 /*yield*/ , runAnimation({
                                    animationDirection: EFFECTS.SHOW.suffix,
                                    effectName: effectName,
                                    effectOptions: __assign(__assign({}, effectDefaultOptions === null || effectDefaultOptions === void 0 ? void 0 : effectDefaultOptions[effectName]), effectOptions),
                                })];
                            case 1:
                                _b.sent();
                                _b.label = 2;
                            case 2:
                                setStyles({
                                    visibility: null
                                });
                                _state.hidden = false;
                                (_a = viewportHandlers === null || viewportHandlers === void 0 ? void 0 : viewportHandlers.onViewPortEnter) === null || _a === void 0 ? void 0 : _a.forEach(function(cb) {
                                    return cb();
                                });
                                return [2 /*return*/ ];
                        }
                    });
                });
            },
            collapse: function() {
                var _a;
                if (!_state.collapsed) {
                    setStyles({
                        visibility: 'hidden !important',
                        height: '0 !important',
                        width: '0 !important',
                        'min-width': '0 !important',
                        overflow: 'hidden !important',
                        'min-height': '0 !important',
                        'margin-bottom': '0 !important',
                    });
                    _state.collapsed = true;
                    if (!_state.hidden) {
                        (_a = viewportHandlers === null || viewportHandlers === void 0 ? void 0 : viewportHandlers.onViewPortLeave) === null || _a === void 0 ? void 0 : _a.forEach(function(cb) {
                            return cb();
                        });
                    }
                }
                return Promise.resolve();
            },
            expand: function() {
                var _a;
                if (_state.collapsed) {
                    setStyles({
                        visibility: _state.hidden ? 'hidden' : null,
                        height: null,
                        width: null,
                        'min-width': null,
                        overflow: null,
                        'min-height': null,
                        'margin-bottom': null,
                    });
                    _state.collapsed = false;
                    if (!_state.hidden) {
                        (_a = viewportHandlers === null || viewportHandlers === void 0 ? void 0 : viewportHandlers.onViewPortEnter) === null || _a === void 0 ? void 0 : _a.forEach(function(cb) {
                            return cb();
                        });
                    }
                }
                return Promise.resolve();
            },
            get collapsed() {
                return _state.collapsed;
            },
            get hidden() {
                return Boolean(_state.hidden);
            },
            get isVisible() {
                if (!metaData.isRendered()) {
                    return false;
                }
                var parentSdk = getSdkInstance();
                while (parentSdk) {
                    if (parentSdk.hidden || parentSdk.collapsed) {
                        return false;
                    }
                    parentSdk = parentSdk.parent;
                }
                return true;
            },
            get isAnimatable() {
                return true;
            },
        };
    };
    return withValidation(_hiddenCollapsedSDKFactory, {
        type: ['object'],
        properties: {
            hide: {
                type: ['function'],
                args: [{
                    type: ['string']
                }, {
                    type: ['object']
                }],
            },
            show: {
                type: ['function'],
                args: [{
                    type: ['string']
                }, {
                    type: ['object']
                }],
            },
        },
    });
};
